loader/src/ui/nodes/ProgressBar.cpp

```diff
@@ -0,0 +1,178 @@
+        m_impl->progressBar->setColor({ 0, 0, 0 });
+        m_impl->progressBar->setOpacity(125);
+        m_impl->progressBar->setZOrder(-1);
+
+        m_impl->progressBarFill = CCSprite::create("GJ_progressBar_001.png");
+        m_impl->progressBarFill->setID("progress-bar-fill");
+        m_impl->progressBarFill->setScale(0.992f);
+        m_impl->progressBarFill->setScaleY(0.86f);
+        m_impl->progressBarFill->setAnchorPoint({ 0, 0.5 });
+        m_impl->progressBarFill->setPosition({ 1.36f, m_impl->progressBar->getScaledContentHeight() / 2.0f });
+        m_impl->progressBarFill->setColor(m_impl->progressBarFillColor);
+        m_impl->progressBarFill->setZOrder(0);
+
+        m_impl->progressBarFillMaxWidth = m_impl->progressBar->getScaledContentWidth();
+        m_impl->progressBarFillMaxHeight = 20.0f;
+
+        m_impl->progressPercentLabel = CCLabelBMFont::create("0%", "bigFont.fnt");
+        m_impl->progressPercentLabel->setID("progress-percent-label");
+        m_impl->progressPercentLabel->setScale(0.5f);
+        m_impl->progressPercentLabel->setAnchorPoint({ 0.5, 0.5 });
+        m_impl->progressPercentLabel->setPosition({ m_impl->progressBar->getScaledContentWidth() / 2.0f, m_impl->progressBar->getScaledContentHeight() / 2.0f });
+        m_impl->progressPercentLabel->setAlignment(CCTextAlignment::kCCTextAlignmentCenter);
+        m_impl->progressPercentLabel->setVisible(m_impl->showProgressPercentLabel);
+        m_impl->progressPercentLabel->setZOrder(1);
+        break;
+    };
+
+    this->setScaledContentSize(m_impl->progressBar->getScaledContentSize());
+
+    m_impl->progressBar->addChild(m_impl->progressBarFill);
+
+    this->addChild(m_impl->progressBar);
+    this->addChild(m_impl->progressPercentLabel);
+
+    this->updateProgress(m_impl->progress);
+};
+
+bool ProgressBar::init() {
+    if (!CCNode::init()) return false;
+
+    this->customSetup();
+
+    return true;
+};
+
+void ProgressBar::setStyle(ProgressBarStyle style) {
+    if (m_impl->style != style) {
+        m_impl->style = style;
+
+        this->removeAllChildren();
+        this->customSetup(); // setup again with new style
+    };
+};
+
+void ProgressBar::setFillColor(ccColor3B color) {
+    m_impl->progressBarFillColor = color;
+    if (m_impl->progressBarFill) m_impl->progressBarFill->setColor(color);
+};
+
+void ProgressBar::updateProgress(float value) {
+    if (value > 100.0f) value = 100.0f;
+    if (value < 0.0f) value = 0.0f;
+
+    m_impl->progress = value;
+
+    if (m_impl->progressBarFill) {
+        float width = m_impl->progressBarFillMaxWidth * (m_impl->progress / 100.0f);
+        m_impl->progressBarFill->setTextureRect({ 0.0f, 0.0f, width, m_impl->progressBarFillMaxHeight });
+    };
+
+    if (m_impl->progressPercentLabel) {
+        auto percentString = fmt::format("{}%", static_cast<int>(m_impl->progress));
```

Comment:
This should probably use `geode::utils::numToString` with a precision of 0. The class should probably also have a way to customize the precision of the percentage